Updated async-await guide for pyo3-asyncio 0.14

awestlake87 opened the pull request:
Hey guys, I'm currently gearing up for a release on `pyo3-asyncio` after merging [pyo3-asyncio PR #30](https://github.com/awestlake87/pyo3-asyncio/pull/30). This is a pretty major change from `v0.13`, but the justification is included within these changes and a migration guide has been created for the pyo3-asyncio README on the `init-fixes` branch.

Let me know if you see any issues with the updated guide or if you have any comments on that PR in general before I release it as pyo3-asyncio 0.14.

davidhewitt commented on guide/src/ecosystem/async-await.md:
```diff
@@ -154,6 +341,360 @@ async fn main() -> PyResult<()> {
+
+`asyncio.get_running_loop` returns the event loop associated with the current OS thread. It can be used inside Python coroutines to spawn concurrent tasks, interact with timers, or in our case signal between Rust and Python. This is all well and good when we are operating on a Python thread, but since Rust threads are not associated with a Python event loop, `asyncio.get_running_loop` will fail w ...
+
+#### The Solution
```
This section is quite long, and goes through two "not quite correct" examples before eventually reaching a set of recommended functions to use.

I feel like users might get lost after copy-pasting either the first or second example and never see the third "recommended" method.

How do you feel about turning this section on its head a bit? Put the recommended APIs and example first, and then offer up a subsection for power users to see `into_future_with_loop` / `scope` etc.

I think that'd let most users see "the solution" immediately, and give advanced users extra material to dive into the nuts and bolts. It might even be nice to put the other "not quite correct" examples somewhere in the PyO3 Asyncio library docs and just link to them from here.

awestlake87 replied:
I see what you mean. The general idea I had for this section was to explain the problems, then provide the intuition behind the new functions in `v0.14` by building to the final recommended solution.

I put this section below the quickstart and general usage because it was intended to get more into the weeds _after_ showing the simpler usage. Maybe this shouldn't even go in the guide though. It might be better served in the API docs with links in the guide. What do you think?

davidhewitt replied:
That seems like a good idea to me. In general I'd prefer to keep this document full of examples of how to get started with PyO3 Asyncio, and let really rich documentation like this for low-level APIs live as close to those APIs as possible (as long as there's ways for users to discover them in the first place).

davidhewitt reviewed (commented):
👍 Very nice improvements! 

I was planning to put out the 0.14.2 release today which was expected to be the last before we get working for 0.15. It would be nice to include this guide, so that users looking at stable docs see the latest PyO3 Asyncio content.

Do you want this guide live today? I could always cherry-pick the current commit and then if there's a few updates that land later in the "main" docs that's ok imo.

Alternatively I don't mind waiting a day or two if you're willing to finish this doc off swiftly 😄 

----

I noticed there's a fair bit of duplication here between this guide and the PyO3 Asyncio README. What's your intention for sharing content between these two documents going forward?

Personally I'd err on keeping this document shorter and more example-focussed with lots of links to PyO3 Asyncio docs. I think it's at risk of diverging from the PyO3 Asyncio codebase over time, which links might help avoid.

awestlake87 commented on guide/src/ecosystem/async-await.md:
```diff
@@ -12,6 +12,176 @@ In the following sections, we'll give a general overview of `pyo3-asyncio` expla
+
+use pyo3::prelude::*;
+
+#[pyo3_asyncio::async_std::main]
+async fn main() -> PyResult<()> {
```
Yes, `pyo3_asyncio::async_std::main` calls `pyo3::prepare_freethreaded_python`. I mentioned that in the migration guide, but maybe it should be in the examples too. 

The reason I did that rather than require `auto-initialize` is because I didn't want to have that feature automatically enabled when people use `pyo3-asyncio` as a dependency (assuming they aren't using the proc macros).

awestlake87 commented on guide/src/ecosystem/async-await.md:
````diff
@@ -12,6 +12,176 @@ In the following sections, we'll give a general overview of `pyo3-asyncio` expla
+
+```
+
+Build your module and rename `libmy_async_module.so` to `my_async_module.so`
+```bash
+cargo build --release && mv target/release/libmy_async_module.so target/release/my_async_module.so
+```
+
+Now, point your `PYTHONPATH` to the directory containing `my_async_module.so`, then you'll be able 
+to import and use it:
+
+```bash
+$ PYTHONPATH=target/release python3
+Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
+[GCC 9.3.0] on linux
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import asyncio
+>>>
+>>> from my_async_module import rust_sleep
+>>> 
+>>> async def main():
+>>>     await rust_sleep()
+>>>
+>>> # should sleep for 1s
+>>> asyncio.run(main())
+>>>
+```
````
Ok, I'll look into it. I've haven't used maturin myself yet, but hopefully I'll get the example working with it sometime today!

awestlake87 commented on guide/src/ecosystem/async-await.md:
````diff
@@ -154,6 +341,360 @@ async fn main() -> PyResult<()> {
+
+You might be surprised to find out that this throws an error:
+```bash
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+RuntimeError: no running event loop
````
I got that traceback from a repl so it's slightly misleading. When I ran it as a file I got this traceback:
```python
Traceback (most recent call last):
  File "example.py", line 5, in <module>
    asyncio.run(rust_sleep())
RuntimeError: no running event loop
```
Was it just the line number that looked odd or is there something else weird about it?

awestlake87 commented on guide/src/ecosystem/async-await.md:
```diff
@@ -154,6 +341,360 @@ async fn main() -> PyResult<()> {
+crate-type = ["cdylib"]
+
+[dependencies]
+pyo3 = { version = "0.14", features = ["extension-module", "auto-initialize"] }
```
You're right it runs fine without it. My bad, I didn't realize that feature was just for applications.

awestlake87 commented:
Alright, I went ahead and applied your suggested changes, shortened the guide, and provided links to the `stable` API docs (which aren't up-to-date on docs.rs just yet) at the end for the sections I removed. BTW, sorry I might have delayed the `0.14.2` release, there's definitely a time lag between us (Central USA vs UK).

I think I can get the suggested changes for the `init-fixes` branch wrapped up within a few hours, but I think as long as I've addressed everything on the guide it should be good to release whenever you want!

davidhewitt commented:
Haha no problem, it's currently 9:20pm here and I'm intending to release in a couple of hours when I get a free moment!
